src/Catalog/Monitoring/Validation/Test/ValidationTestRunner.cs

```diff
@@ -0,0 +1,129 @@
+        /// Returns whether or not the validation test should run on a <see cref="ValidationTestRunData"/>.
+        /// </summary>
+        Task<bool> ShouldRun(ValidationTestRunData data);
+
+        /// <summary>
+        /// Validates a package.
+        /// </summary>
+        /// <returns>A <see cref="ValidationTestResult"/> which contains the results of the validation.</returns>
+        Task<ValidationTestResult> Run(ValidationTestRunData data);
+    }
+    
+    /// <summary>
+    /// Abstract class with the shared functionality between all <see cref="IValidationTestRunner"/> implementations.
+    /// </summary>
+    public abstract class ValidationTestRunner : IValidationTestRunner
+    {
+        /// <summary>
+        /// The number of tries attempted before the test is considered to have failed.
+        /// </summary>
+        public const int MaxTries = 3;
+
+        /// <summary>
+        /// Constructor for testing purposes.
+        /// </summary>
+        protected ValidationTestRunner()
+        {
+        }
+
+        protected ValidationTestRunner(
+            IDictionary<FeedType, SourceRepository> feedToSource)
+        {
+            _timestampMetadataResourceV2 = feedToSource[FeedType.HttpV2].GetResource<IPackageTimestampMetadataResource>();
+        }
+
+        private readonly IPackageTimestampMetadataResource _timestampMetadataResourceV2;
+
+        /// <summary>
+        /// Checks that the current batch of catalog entries contains the entry that was created from the current state of the V2 feed.
+        /// </summary>
+        public virtual async Task<bool> ShouldRun(ValidationTestRunData data)
+        {
+            var timestampV2 = await _timestampMetadataResourceV2.Get(data);
+            var timestampCatalog = await PackageTimestampMetadata.FromCatalogEntries(data.Client, data.Entries);
+            
+            if (!timestampV2.Last.HasValue)
+            {
+                throw new ValidationTestException("Cannot get timestamp data for package from the V2 feed!");
+            }
+
+            if (!timestampCatalog.Last.HasValue)
+            {
+                throw new ValidationTestException("Cannot get timestamp data for package from the catalog!");
+            }
+
+            if (timestampCatalog.Last > timestampV2.Last)
+            {
+                throw new ValidationTestException("The timestamp in the catalog is newer than the timestamp in the feed! This should never happen because all data flows from the catalog into the feed!");
+            }
+
+            // If the timestamp metadata in the catalog is LESS than that of the feed, we must not be looking at the latest entry that corresponds with this package, so skip the test for now.
+            // If the timestamp metadata in the catalog is EQUAL to that of the feed, we are looking at the latest catalog entry that corresponds with this package, so run the test.
+            return timestampCatalog.Last == timestampV2.Last;
```

Comment:
Comparing timestamps this way is risky.. there are precision related concerns, etc.